aiida/orm/implementation/sqlalchemy/utils.py

```diff
@@ -80,11 +80,12 @@ def __setattr__(self, key, value):
             self._flush(fields=fields)
 
     def is_saved(self):
-        """Retun whether the wrapped model instance is saved in the database.
+        """Return whether the wrapped model instance is saved in the database.
 
         :return: boolean, True if the model is saved in the database, False otherwise
         """
-        return self._model.id is not None
+        with self._model.session.no_autoflush:
```

Comment:
Why would the session be flushed when we request an attribute of a model instance?

Reply:
If you have already tried to (incorrectly) set an attribute on the model.

This failed in `tests/orm/test_groups.py::TestGroups::test_rename_existing`

```python
        with self.assertRaises(exceptions.IntegrityError):
            group_b.label = label_group_a
```

Instead of being caught below, by `save()` and being converted to an aiida `IntegrityError`, it was earlier failing here

Reply:
How was it failing though? Does this have to do with the `ModelWrapper` we add? I though flushing would only be performed when pushing data from memory to database and not the reverse. So why would setting some attribute on the model (incorrect or not) lead to the `self._model.id` call resulting in a flush?

Reply:
Oops just removed again in f8ea014, because I thought I could no longer reproduce, but then realised I wasn't using the SQLA backend. So you should see it in the results of that CI run, but basically without it, this is the error you get:

```
py37-sqla run-test: commands[0] | pytest tests/orm/test_groups.py::TestGroups::test_rename_existing
==================================================================================== test session starts ====================================================================================
platform darwin -- Python 3.7.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
cachedir: .tox/py37-sqla/.pytest_cache
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/chrisjsewell/Documents/GitHub/aiida-core-origin, configfile: pyproject.toml
plugins: benchmark-3.4.1, datadir-1.3.1, asyncio-0.15.1, rerunfailures-9.1.1, regressions-2.2.0, cov-2.10.1, timeout-1.4.2
collected 1 item                                                                                                                                                                            

tests/orm/test_groups.py FE                                                                                                                                                           [100%]

========================================================================================== ERRORS ===========================================================================================
___________________________________________________________________ ERROR at teardown of TestGroups.test_rename_existing ____________________________________________________________________

cls = <class 'tests.orm.test_groups.TestGroups'>

    @classmethod
    def tearDownClass(cls):
        """Tear down test class.
    
        Note: Also cleans file repository.
        """
        # Double check for double security to avoid to run the tearDown
        # if this is not a test profile
    
        check_if_tests_can_run()
        if orm.autogroup.CURRENT_AUTOGROUP is not None:
            orm.autogroup.CURRENT_AUTOGROUP.clear_group_cache()
>       cls.clean_db()

aiida/backends/testbase.py:96: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
aiida/backends/testbase.py:131: in clean_db
    cls.__backend_instance.clean_db()
aiida/backends/sqlalchemy/testbase.py:36: in clean_db
    with self.backend.transaction() as session:
.tox/py37-sqla/lib/python3.7/contextlib.py:112: in __enter__
    return next(self.gen)
aiida/orm/implementation/sqlalchemy/backend.py:88: in transaction
    session.begin_nested()
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:1365: in begin_nested
    return self.begin(nested=True)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/util/deprecations.py:298: in warned
    return fn(*args, **kwargs)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:1320: in begin
    trans = self._transaction._begin(nested=nested)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:629: in _begin
    self._assert_active()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlalchemy.orm.session.SessionTransaction object at 0x7f9bd3474ad0>, prepared_ok = False, rollback_ok = False, deactive_ok = False, closed_msg = 'This transaction is closed'

    def _assert_active(
        self,
        prepared_ok=False,
        rollback_ok=False,
        deactive_ok=False,
        closed_msg="This transaction is closed",
    ):
        if self._state is COMMITTED:
            raise sa_exc.InvalidRequestError(
                "This session is in 'committed' state; no further "
                "SQL can be emitted within this transaction."
            )
        elif self._state is PREPARED:
            if not prepared_ok:
                raise sa_exc.InvalidRequestError(
                    "This session is in 'prepared' state; no further "
                    "SQL can be emitted within this transaction."
                )
        elif self._state is DEACTIVE:
            if not deactive_ok and not rollback_ok:
                if self._rollback_exception:
                    raise sa_exc.PendingRollbackError(
                        "This Session's transaction has been rolled back "
                        "due to a previous exception during flush."
                        " To begin a new transaction with this Session, "
                        "first issue Session.rollback()."
                        " Original exception was: %s"
                        % self._rollback_exception,
>                       code="7s2a",
E                       sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
E                       (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "db_dbgroup_label_type_string_key"
E                       DETAIL:  Key (label, type_string)=(group_a, core) already exists.
E                       
E                       [SQL: UPDATE db_dbgroup SET label=%(label)s WHERE db_dbgroup.id = %(db_dbgroup_id)s]
E                       [parameters: {'label': 'group_a', 'db_dbgroup_id': 3}]
E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)

.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:608: PendingRollbackError
------------------------------------------------------------------------------------ Captured log setup -------------------------------------------------------------------------------------
DEBUG    pgsu:__init__.py:141 Trying to connect via "psycopg2"...
========================================================================================= FAILURES ==========================================================================================
______________________________________________________________________________ TestGroups.test_rename_existing ______________________________________________________________________________

    def _execute_context(
        self,
        dialect,
        constructor,
        statement,
        parameters,
        execution_options,
        *args,
        **kw
    ):
        """Create an :class:`.ExecutionContext` and execute, returning
        a :class:`_engine.CursorResult`."""
    
        branched = self
        if self.__branch_from:
            # if this is a "branched" connection, do everything in terms
            # of the "root" connection, *except* for .close(), which is
            # the only feature that branching provides
            self = self.__branch_from
    
        try:
            conn = self._dbapi_connection
            if conn is None:
                conn = self._revalidate_connection()
    
            context = constructor(
                dialect, self, conn, execution_options, *args, **kw
            )
        except (exc.PendingRollbackError, exc.ResourceClosedError):
            raise
        except BaseException as e:
            self._handle_dbapi_exception(
                e, util.text_type(statement), parameters, None, None
            )
    
        if (
            self._transaction
            and not self._transaction.is_active
            or (
                self._nested_transaction
                and not self._nested_transaction.is_active
            )
        ):
            self._invalid_transaction()
    
        elif self._trans_context_manager:
            TransactionalContext._trans_ctx_check(self)
    
        if self._is_future and self._transaction is None:
            self._autobegin()
    
        context.pre_exec()
    
        if dialect.use_setinputsizes:
            context._set_input_sizes()
    
        cursor, statement, parameters = (
            context.cursor,
            context.statement,
            context.parameters,
        )
    
        if not context.executemany:
            parameters = parameters[0]
    
        if self._has_events or self.engine._has_events:
            for fn in self.dispatch.before_cursor_execute:
                statement, parameters = fn(
                    self,
                    cursor,
                    statement,
                    parameters,
                    context,
                    context.executemany,
                )
    
        if self._echo:
    
            self._log_info(statement)
    
            stats = context._get_cache_stats()
    
            if not self.engine.hide_parameters:
                self._log_info(
                    "[%s] %r",
                    stats,
                    sql_util._repr_params(
                        parameters, batches=10, ismulti=context.executemany
                    ),
                )
            else:
                self._log_info(
                    "[%s] [SQL parameters hidden due to hide_parameters=True]"
                    % (stats,)
                )
    
        evt_handled = False
        try:
            if context.executemany:
                if self.dialect._has_events:
                    for fn in self.dialect.dispatch.do_executemany:
                        if fn(cursor, statement, parameters, context):
                            evt_handled = True
                            break
                if not evt_handled:
                    self.dialect.do_executemany(
                        cursor, statement, parameters, context
                    )
            elif not parameters and context.no_parameters:
                if self.dialect._has_events:
                    for fn in self.dialect.dispatch.do_execute_no_params:
                        if fn(cursor, statement, context):
                            evt_handled = True
                            break
                if not evt_handled:
                    self.dialect.do_execute_no_params(
                        cursor, statement, context
                    )
            else:
                if self.dialect._has_events:
                    for fn in self.dialect.dispatch.do_execute:
                        if fn(cursor, statement, parameters, context):
                            evt_handled = True
                            break
                if not evt_handled:
                    self.dialect.do_execute(
>                       cursor, statement, parameters, context
                    )

.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1772: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "db_dbgroup_label_type_string_key"
E       DETAIL:  Key (label, type_string)=(group_a, core) already exists.

.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/default.py:717: UniqueViolation

The above exception was the direct cause of the following exception:

self = <tests.orm.test_groups.TestGroups testMethod=test_rename_existing>

    def test_rename_existing(self):
        """Test that renaming to an already existing name is not permitted."""
        label_group_a = 'group_a'
        label_group_b = 'group_b'
    
        orm.Group(label=label_group_a, description='I am the Original G').store()
    
        # Before storing everything should be fine
        group_b = orm.Group(label=label_group_a, description='They will try to rename me')
    
        # Storing for duplicate group name should trigger UniquenessError
        with self.assertRaises(exceptions.IntegrityError):
            group_b.store()
    
        # Reverting to unique name before storing
        group_b.label = label_group_b
        group_b.store()
    
        # After storing name change to existing should raise
        with self.assertRaises(exceptions.IntegrityError):
>           group_b.label = label_group_a

tests/orm/test_groups.py:255: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
aiida/orm/groups.py:177: in label
    self._backend_entity.label = label
aiida/orm/implementation/sqlalchemy/groups.py:67: in label
    self._dbmodel.label = label
aiida/orm/implementation/sqlalchemy/utils.py:78: in __setattr__
    if self.is_saved() and self._is_mutable_model_field(key):
aiida/orm/implementation/sqlalchemy/utils.py:87: in is_saved
    return self._model.id is not None
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py:481: in __get__
    return self.impl.get(state, dict_)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py:926: in get
    value = self._fire_loader_callables(state, key, passive)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py:957: in _fire_loader_callables
    return state._load_expired(state, passive)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/state.py:710: in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/loading.py:1483: in load_scalar_attributes
    no_autoflush=no_autoflush,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/loading.py:411: in load_on_ident
    execution_options=execution_options,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/loading.py:561: in load_on_pk_identity
    bind_arguments=bind_arguments,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:1643: in execute
    _parent_execute_state is not None,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/context.py:290: in orm_pre_session_exec
    session._autoflush()
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:2204: in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/util/compat.py:207: in raise_
    raise exception
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:2193: in _autoflush
    self.flush()
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3298: in flush
    self._flush(objects)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3438: in _flush
    transaction.rollback(_capture_exception=True)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py:72: in __exit__
    with_traceback=exc_tb,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/util/compat.py:207: in raise_
    raise exception
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3398: in _flush
    flush_context.execute()
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py:456: in execute
    rec.execute(self)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py:633: in execute
    uow,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py:239: in save_obj
    update,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py:999: in _emit_update_statements
    statement, multiparams, execution_options=execution_options
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1583: in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/sql/elements.py:324: in _execute_on_connection
    self, multiparams, params, execution_options
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1462: in _execute_clauseelement
    cache_hit=cache_hit,
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1815: in _execute_context
    e, statement, parameters, cursor, context
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1996: in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/util/compat.py:207: in raise_
    raise exception
.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/base.py:1772: in _execute_context
    cursor, statement, parameters, context
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
E       (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "db_dbgroup_label_type_string_key"
E       DETAIL:  Key (label, type_string)=(group_a, core) already exists.
E       
E       [SQL: UPDATE db_dbgroup SET label=%(label)s WHERE db_dbgroup.id = %(db_dbgroup_id)s]
E       [parameters: {'label': 'group_a', 'db_dbgroup_id': 3}]
E       (Background on this error at: https://sqlalche.me/e/14/gkpj)

.tox/py37-sqla/lib/python3.7/site-packages/sqlalchemy/engine/default.py:717: IntegrityError
```

Reply:
Yeah, but I think that the solution should be different. The test is _expecting_ to raise, specifically, it is expecting an exception related to a uniqueness error because it is trying to store a group with a label that already exists. I think what changed is simply the type of exception that is being thrown. It is checking for a `exceptions.IntegrityError` but now a `UniqueViolation` seems to be thrown. I think we should simply change the expected exception. And maybe understand why the exception has changed, but if that is to be expected after the upgrade, than that would make sense.

Reply:
Yeh but it is not expecting to raise in this part of the code; `is_saved()` should never except,
and it is not now raising a `UniqueViolation`; it is raising an `sqlalchemy.exc.IntegrityError` rather than an `aiida.exceptions.Integrity`, and this is directly because it is excepting in `is_saved()`, rather than returning `False`, then failing later in `save()`, where the exception gets "swapped":

```python
        try:
            commit = not self._in_transaction()
            self._model.save(commit=commit)
        except IntegrityError as exception:
            self._model.session.rollback()
            raise exceptions.IntegrityError(str(exception))
```

This is definitely the correct solution

Reply:
it basically goes `__setattr__` -> `is_saved` (False)-> `_flush` -> `save`, and should only except when it gets to `save`

Reply:
> is_saved() should never except, and it is not now raising a UniqueViolation

That is my point though. I realize this is not your doing but `sqlalchemy`'s design, but to me it seems really weird that simply accessing a models attribute causes the session to be flushed. I am sure they have a reasoning that I am not seeing but clearly this results in weird behavior. Fine to keep the change but please add a comment why this `no_autoflush` is needed here because it is not at all obvious

Reply:
added comment

>  realize this is not your doing but sqlalchemy's design, but to me it seems really weird that simply accessing a models attribute causes the session to be flushed.

well this is likely more to do with our overrides of the `__getattr__`, `__setattr__` methods, which force this behaviour; remember this is all actually happening during the (incorrect) setting of an attribute